source/common/network/connection_impl.cc

```diff
@@ -34,6 +34,16 @@ Address::InstanceConstSharedPtr getNullLocalAddress(const Address::Instance& add
 }
 } // namespace
 
+int ConnectionImplUtility::createSocket(Address::InstanceConstSharedPtr address,
+                                        Optional<Address::InstanceConstSharedPtr> source_address) {
+  int fd = address->socket(Address::SocketType::Stream);
+  if (fd >= 0 && source_address.valid()) {
+    int rc = source_address.value()->bind(fd);
+    ASSERT(rc >= 0);
```

Comment:
Lo, the rampaging herd of yaks.

So my clever idea of returning -2 for bind failures and treating that as an immediate disconnect in connection_impl.cc fails because the caller doesn't have callbacks registered yet so you simply can't disconnect immediately.
We can't ask for an async callback because we don't have a file descriptor to do that with.    I tried then just notifying callers when they got around to addCallback which works in unit tests but fails in integration tests where some various components assert they have a viable connection when it's created.

So we either have to bind in the constructor, and when it fails leave the fd open but register for immediate disconnect, or we have to check at all of the 15 call sites where connections are created to ensure it's valid.  Yuck.  I'm going to leave this as a TODO for today and take a whack at it later.